src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+    };
+}
+
+// Need to reproduce the IToken interface from vscode-textmate due to
+// the odd way it has to be required
+// https://github.com/Microsoft/vscode-textmate/blob/46af9487e1c8fa78aa1f2e2/release/main.d.ts#L161-L165
+export interface IToken {
+    startIndex: number;
+    readonly endIndex: number;
+    readonly scopes: string[];
+}
+
+// The different types of matched textmate tokens.
+// Mirrors the enum at https://code.visualstudio.com/docs/extensionAPI/vscode-api#FoldingRangeKind
+enum MatchType {
+    Comment = 1,
+    Imports = 2,
+    Region = 3,
+}
+
+// Defines a pair of tokens (start, end) or pair or line numbers (startLine, endLine) which define
+// a folding range in source code.
+class MatchedToken {
+    // Start IToken object of the range.  Optional if startLine is specified
+    public start: IToken;
+    // The starting zero-based line number of the range.
+    public startline: number;
+    // End IToken object of the range.  Optional if startLine is specified
+    public end: IToken;
+    // The ending zero-based line number of the range.
+    public endline: number;
+    // The type of range this is
+    public matchType: MatchType;
+
+    constructor(start: IToken = null,
+                end: IToken = null,
+                startLine: number = null,
+                endLine: number = null,
+                matchType: MatchType,
+                document: vscode.TextDocument) {
+        this.start = start;
+        this.end = end;
+        if (startLine === null) {
+            this.startline = document.positionAt(start.startIndex).line;
+        } else {
+            this.startline = startLine;
+        }
+        if (endLine === null) {
+            this.endline = document.positionAt(end.startIndex).line;
+        } else {
+            this.endline = endLine;
+        }
+        this.matchType = matchType;
+    }
+
+    // Returns whether this pair of matched tokens/line numbers, is a valid folding range
+    public isValidRange(): boolean {
+        // Folding ranges must span at least 2 lines
+        return (this.endline - this.startline >= 1);
+    }
+
+    // Creates a vscode.FoldingRange object based on this object
+    public toFoldingRange(): vscode.FoldingRange {
+        let rk: vscode.FoldingRangeKind = vscode.FoldingRangeKind.Region;
+        switch (this.matchType) {
+            case MatchType.Comment: { rk = vscode.FoldingRangeKind.Comment; break; }
+            case MatchType.Imports: { rk = vscode.FoldingRangeKind.Imports; break; }
+        }
+        return new vscode.FoldingRange(this.startline, this.endline, rk);
+    }
+}
+
+interface IMatchedTokenList extends Array<MatchedToken> {}
+
+export class FoldingProvider implements vscode.FoldingRangeProvider {
+    private powershellGrammar;
+
+    public async provideFoldingRanges(
+        document: vscode.TextDocument,
+        context: vscode.FoldingContext,
+        token: vscode.CancellationToken,
+    ): Promise<vscode.FoldingRange[]> {
```

Comment:
nitpick: I noticed some of your functions look like this:

```typescript
public foo(
    doc: bar
    context: baz
) : Boo {
// ...
}
```

and some look like:

```typescript
public foo(doc: bar
           context: baz) : Boo {
// ...
}
```

Is there a reason for the different indenting? Can we align those if not? (I personally prefer the first style 😄)

Reply:
It's my ruby-isms coming through :-).  Will reformat to first one :-)

Reply:
Somewhat fixed.  I've reformatted the code I wrote,  haven't done the code I copy-pasta'd from other sources.